### PR TITLE
eth: make traceChain avoid OOM on long-running tracing

### PR DESCRIPTION
**Pull request by holiman**
This PR changes long-running chain tracing, so that it at some points releases the memory trie db, and switch over to a fresh disk-backed trie.

### CODE REVIEW

**Comment by holiman** [comment]
Finally got around to testing this on a node.. 
```
> debug.getAccessibleState(0,"latest")
13401259
> debug.getAccessibleState(13401260,"latest")
13414706
```
So if tracing starts at `13401259`, then it will have to chug through `13414706-13401259`=`13447` blocks before the memory is cleared. Should be doable. 
```
INFO [11-02|17:40:22.208] Tracing chain segment                    start=13,401,259 end=13,537,532 current=13,414,693 transactions=2,545,920 elapsed=4h3m22.508541811s
INFO [11-02|17:40:30.543] Tracing chain segment                    start=13,401,259 end=13,537,532 current=13,414,700 transactions=2,547,143 elapsed=4h3m30.843300651s
INFO [11-02|17:40:35.864] Deep froze chain segment                 blocks=3  elapsed=36.277ms           number=13,448,639 hash=c57e2d..edd170
INFO [11-02|17:40:39.146] Tracing chain segment                    start=13,401,259 end=13,537,532 current=13,414,706 transactions=2,548,254 elapsed=4h3m39.446252969s
INFO [11-02|17:40:39.578] Found disk backend for state trie 
```
Yay!
```
INFO [11-02|17:40:41.366] Downloader queue stats                   receiptTasks=0 blockTasks=0 itemSize=69.51KiB throttle=3772
WARN [11-02|17:40:43.371] Tracing failed                           hash=dff138..a8826e block=13,414,708 err="tracing failed: nonce too high: address 0xAe45a8240147E6179ec7c9f92c5A18F9a97B3fCA, tx: 680930 state: 680929"
WARN [11-02|17:40:46.235] Chain tracing failed                     start=13,401,259 end=13,537,532 transactions=2,548,566 elapsed=4h3m46.535467752s  err="processing block 13414708 failed: could not apply tx 71 [0xdff1382a7bd005a5e54425f65e937b8d477dfd31c21a78e71a5781f873a8826e]: nonce too high: address 0xAe45a8240147E6179ec7c9f92c5A18F9a97B3fCA, tx: 680930 state: 680929"
```
Nooo!

---- 

Memory increase during those 13k block tracing: 
Screenshot 2021-11-02 at 18-24-30 Single Geth - Grafana

**Comment by holiman** [comment]
Yup, now this works. 
Screenshot 2021-11-02 at 22-50-35 Single Geth - Grafana
The drop there is when it finds a disk trie: 
```
INFO [11-02|17:40:46.488] Imported new chain segment               blocks=1  txs=481  mgas=29.983  elapsed=464.753ms          mgasps=64.514  number=13,538,640 hash=f11200..e24c65 dirty=689.33MiB
INFO [11-02|17:40:46.490] Unindexed transactions                   blocks=1  txs=97   tail=13,538,541 elapsed=1.890ms
INFO [11-02|17:41:06.013] Imported new chain segment               blocks=1  txs=498  mgas=29.541  elapsed=231.347ms          mgasps=127.693 number=13,538,641 hash=489220..d81a9f dirty=690.82MiB
INFO [11-02|17:41:06.017] Unindexed transactions                   blocks=1  txs=172  tail=13,538,542 elapsed=3.511ms
...
INFO [11-02|22:47:21.454] Tracing chain segment                    start=13,401,259 end=13,538,940 current=13,414,701 transactions=2,547,632 elapsed=3h59m24.619355258s
INFO [11-02|22:47:32.152] Tracing chain segment                    start=13,401,259 end=13,538,940 current=13,414,706 transactions=2,548,254 elapsed=3h59m35.316871285s
INFO [11-02|22:47:32.155] Found disk backend for state trie 
INFO [11-02|22:47:40.320] Tracing chain segment                    start=13,401,259 end=13,538,940 current=13,414,710 transactions=2,548,816 elapsed=3h59m43.485377022s
INFO [11-02|22:47:44.500] Switching to prefer-disk mode for tracing size=11.78MiB
INFO [11-02|22:47:49.864] Tracing chain segment                    start=13,401,259 end=13,538,940 current=13,414,717 transactions=2,550,574 elapsed=3h59m53.028526974s
```

**Comment by holiman** [comment]
Rebased, lightly squashed, fixed up. PTAL

**Comment by holiman** [comment]
Has been chugging along the entire night (started a few minutes after midnight ), and switches over from time to time: 
Screenshot 2021-11-03 at 10-11-52 Single Geth - Grafana

**Review by s1na — APPROVED** [review]
LGTM. Also tested on a node

File: eth/state_accessor.go [hunk 1]
```diff
@@ -35,7 +35,17 @@ import (
 // are attempted to be reexecuted to generate the desired state. The optional
 // base layer statedb can be passed then it's regarded as the statedb of the
 // parent block.
-func (eth *Ethereum) stateAtBlock(block *types.Block, reexec uint64, base *state.StateDB, checkLive bool) (statedb *state.StateDB, err error) {
+// Parameters:
+// - block: The block for which we want the state (== state at the stateRoot of the parent)
```
Comment (s1na) [inline eth/state_accessor.go:39]: @holiman `== state at the stateRoot of the parent` I'm a bit confused by this comment. I feel like what this function does is give you the state at the stateRoot of the block itself?

Comment (holiman) [reply eth/state_accessor.go:39]: This method returns the `block.Root` yes. 
So if you want to execute block `N`, you need the `stateAt( N-1)`. 
So you need to call this method with `N-1`. 
So the comment should have said

``
// - block: the block to look up the stateRoot of. This needs to be the parent of the block you want to execute
```

Comment (s1na) [reply eth/state_accessor.go:39]: Ah gotcha thanks

**Pull request merged**